wasmcloud-chat/actors/messages/src/lib.rs

```
@@ -1,9 +1,172 @@
-}
\ No newline at end of file
+    messages::Handlers::register_process_message(process_message);
+}
+
+// Processing a message:
+// 1 - validate the message
+// 2 - publish it to the appropriate message broker subject
+// 3 - emit event to event stream
+// 4 - ACK
+fn process_message(message: ChannelMessage) -> HandlerResult<ProcessAck> {
+    let target = MessageTarget::from(&message);
+    if let MessageTarget::Unknown(message) = target {
+        return Err(format!("Unable to select target for message: {}", message).into());
+    }
+
+    Ok(validate_message(&message)
+        .and_then(|_| publish_broker_message(&target, &message))
+        .and_then(|_| emit_stream_event(&target, &message))
+        .map_or_else(
+            |_| ProcessAck::success(&message.message_id),
+            |_| {
+                ProcessAck::fail(
+                    &message.message_id,
+                    "Did not publish and emit to event stream",
+                )
+            },
+        ))
+}
+
+fn validate_message(message: &messages::ChannelMessage) -> HandlerResult<()> {
+    if message.message_id.is_empty() {
+        return Err("No message ID".into());
+    }
+    if message.target_url == format!("wcchat://users/{}", message.origin_user_id) {
+        return Err("Cannot send messages to yourself".into());
+    }
+    if message.message_text.is_empty() {
+        return Err("Cannot send empty messages".into());
+    }
+    if !VALID_MESSAGE_TYPES
+        .to_vec()
+        .contains(&message.message_type.as_str())
+    {
+        return Err("Invalid/unrecognized message type".into());
+    }
+    Ok(())
+}
+
+fn publish_broker_message(
+    target: &MessageTarget,
+    message: &messages::ChannelMessage,
+) -> HandlerResult<()> {
+    let topic = match target {
+        MessageTarget::User(s) => format!("wcc.events.user.{}", s),
+        MessageTarget::Room(s) => format!("wcc.events.room.{}", s),
+        _ => "".to_string(),
+    };
```

It might be better to return an error here mentioning that there was no user or room available for that target, unless we want to explicitly send the unknown message targets to a specific topic

---

The target validation has already taken place prior to this, so we're only using the empty string here to satisfy the "must match all variants" requirement of matching on an enum.

We could also certainly add error checking elsewhere that prevents publication to an empty topic. I tried a variant of this that had `Result`s everywhere and it felt super ugly. Again, personal preference.

---

Also TIL you can add a review comment to more than one line of code at a time :)

---

As long as that is essentially "unreachable" code due to our code path, this looks fine to me 👍🏻